### PR TITLE
Added two possibilities in Mod( )

### LINKED ISSUE #13543: Mod(Mod(x + 1, 2) + 1, 2) should simplify to Mod(x, 2)
From [stackoverflow](https://stackoverflow.com/questions/46914006/modulo-computations-in-sympy-fail)

Also, something like `Mod(foo*Mod(x + 1, 2) + non_mod_terms + 1, 2)` could be simplified. Recursively.

### PR DESCRIPTION
**Pull request by ashishkg0022**
case where `(A mod c + B)mod c` can be simplified was missing.
 ` Mod(Mod(A, c) + B, c) == Mod(A + B, c)`
used the above concept to add a possibilty which was missing
Fixes #13543 

## This PR changes how Mod autosimplifies arguments
m: Mod is now distributed over Add and Mul args to provide additional simplification

### CODE REVIEW

**Comment by ashishkg0022** [comment]
ping @smichr

File: sympy/core/mod.py [hunk 1]
```diff
@@ -107,6 +107,17 @@ def doit(p, q):
             elif (qinner*(q + qinner)).is_nonpositive:
                 # |qinner| < |q| and have different sign
                 return p
+        elif any(isinstance(arg, cls) for arg in p.args):
```
Comment (asmeurer) [inline sympy/core/mod.py]: Not all operations commute with mod. There needs to be a check on the type of `p` here.

File: sympy/core/mod.py [hunk 2]
```diff
@@ -107,6 +107,17 @@ def doit(p, q):
...
+            #separating into modulus and non modulus
+            both_l = non_mod_l, mod_l = [], []
+            for arg in p.args:
+                both_l[isinstance(arg,cls)].append(arg)
+            #if q same for all
+            if all(inner.args[1] == q for inner in mod_l) and \
+                    all(cls(z,q) == z for z in non_mod_l):
```
Comment (asmeurer) [inline sympy/core/mod.py]: I don't think this last check is necessary. Mod(Mod(A, c) + B, c) == Mod(A + B, c) unconditionally.

**Comment by asmeurer** [comment]
The reason this works is that Mod distributes over addition and is idempotent (with the same modulus), so 

Mod(Mod(x + 1, 2) + 1 , 2) == Mod(Mod(x + 1, 2), 2) + Mod(1, 2) == Mod(x + 1, 2) + Mod(1, 2) == Mod(x + 1 + 1, 2).

The same would be true over multiplication.

**Comment by ashishkg0022** [comment]
@asmeurer , Is this the right way to solve the issue or it needs something different?

**Comment by ashishkg0022** [comment]
I have made the required changes. please see it.

**Comment by ashishkg0022** [comment]
ping @asmeurer

File: sympy/core/mod.py [hunk 3]
```diff
@@ -107,6 +107,16 @@ def doit(p, q):
             elif (qinner*(q + qinner)).is_nonpositive:
                 # |qinner| < |q| and have different sign
                 return p
+        elif isinstance(p, Add):
+            #separating into modulus and non modulus
+            both_l = non_mod_l, mod_l = [], []
```
Comment (smichr) [inline sympy/core/mod.py]: 🥇 OK, this is a cool approach to binary sifting! I've copied it to a couple of places in #13599. Is this your own or did you learn this from somewhere else?

Comment (ashishkg0022) [reply sympy/core/mod.py]: @smichr  thanks, I learnt it from a comment on a query in StackOverflow.

**Comment by smichr** [comment]
From what Aaron is saying, I suspect that one could do this:

```
Mod(Mod(x+2,4)*(x+4),4) -> Mod⁡((x+4)Mod⁡(x+2,4),4)
Mod(Mod(x+2,4)*4,4) -> Mod(x+2,4)*Mod(4,4) -> 0
```

So I *think* you would just add a loop for Mul and use prod instead of sum.

**Comment by ylemkimon** [comment]
I'm not sure, but doesn't commutativity affects the result?

**Comment by ashishkg0022** [comment]
ping @smichr

**Comment by smichr** [comment]
Just some minor touch-up. Otherwise this looks good to me.

**Comment by ashishkg0022** [comment]
thanks..

**Pull request merged**